### ISSUE TITLE
Nested object properties

### ISSUE THREAD

**Issue by sollecitom** [uchuhimo/konf#14]
Hey @uchuhimo thanks again for your great work here. It would be great to be able to specify `required` and `optional` nested object properties.

At the moment, this is technically possible, but a bit awkward, requiring to use a `Map<String, Any>`, and then using that to create a `Config`, perhaps with one or more `Spec`s.

Ideally, the following should work:

```
val outer = object : ConfigSpec() {
    
     val inner = object : ConfigSpec("address") {
            
            val host by required<String>()
            val port by required<Int>()
     }
     
     val settings by required<Config>().withSchema(inner)
}

val configuration = Config.invoke { addSpec(outer) }.from.map.kv(mapOf("settings.address.host" to "127.0.0.1", "settings.address.port" to 8080))

val settings = configuration[outer.settings]
val host = settings[outer.inner.host]
```

Not sure if there's a better way of doing the above, perhaps by using multiple `Spec`s.

**Comment by sollecitom** [comment]
Alternatively, perhaps more simple to implement, a `fun <TYPE, MAPPED> map(convert: value: TYPE -> MAPPED): Item<MAPPED` from `Item<TYPE>`.

So that `val nested by required<Map<String, Any>().map { Config.invoke().from.map.kv(it).also { it.addSpec(NestedSpec) } }` could work.

**Comment by uchuhimo** [comment]
@sollecitom Thank you for your RFE (Request for Enhancement)! I have added support for nested config spec and spec prefix inference in Konf. Now you can organize your spec in the following way:

```kotlin
object Service : ConfigSpec() {
    val name by optional("test")

    object Backend : ConfigSpec() {
        val host by optional("127.0.0.1")
        val port by optional(7777)

        object Login : ConfigSpec() {
            val user by optional("admin")
            val password by optional("123456")
        }
    }

    object UI : ConfigSpec() {
        val host by optional("127.0.0.1")
        val port by optional(8888)
    }
}
```

Konf will be aware of their relationship and inference their prefix from their object name, so you can use this spec in the following way:

```kotlin
val config = Config { addSpec(Service) }
val user = config[Service.Backend.Login.user]
val uiPort = config<Int>("service.ui.host")
```

You can try this feature by using master snapshot from JitPack. I will release this feature in the next version.

Is this feature what you want? Or any missing part to support your use case?

**Comment by sollecitom** [comment]
Hi again, thanks for your quick response. I was keen on having a
sub-section of the configuration returned as Config with schema. So
config[prop] returning a Config set up with the schema specified at
property creation time.

**Comment by uchuhimo** [comment]
You can use `Config#at(String)` to get sub-section of a config as a config. The example you use can be written as:

```kotlin
object outer : ConfigSpec("settings") {
     object inner : ConfigSpec("address") {
            val host by required<String>()
            val port by required<Int>()
     }
}

val configuration = Config { addSpec(outer) }.from.map.kv(mapOf("settings.address.host" to "127.0.0.1", "settings.address.port" to 8080))

val settings: Config = configuration.at("settings")
val host = settings[outer.inner.host]
```

You can also use `val settings = configuration.at(outer.prefix)` instead of `val settings = configuration.at("settings")` to avoid repeating the prefix.

Please correct me if I misunderstand your use case.

**Comment by grambbledook** [comment]
> You can use `Config#at(String)` to get sub-section of a config as a config. The example you use can be written as:

Hi, I've just tries the code from your example:

```
Exception in thread "main" com.uchuhimo.konf.NoSuchItemException: cannot find item host in config
	at com.uchuhimo.konf.BaseConfig.getOrNull(BaseConfig.kt:170)
	at com.uchuhimo.konf.BaseConfig.getOrNull(BaseConfig.kt:167)
	at com.uchuhimo.konf.BaseConfig.getOrNull(BaseConfig.kt:167)
	at com.uchuhimo.konf.BaseConfig.getOrNull$default(BaseConfig.kt:117)
	at com.uchuhimo.konf.BaseConfig.get(BaseConfig.kt:106)
	at com.foo.bar.MainKt.main(Main.kt:28)
```

konf version: 0.12
kotlin version:  1.3.10

**Comment by uchuhimo** [comment]
@grambbledook Thank you for your feedback. I have not been aware that the feature of nested config spec is not included in v0.12 until today. The code cannot work under v0.12. I have launched a new release v0.13, which includes all the necessary features to run the example. You can try it out. Apologize for my mistake.
